### PR TITLE
[Ruby][Rails] Syntax improvements

### PR DESCRIPTION
**Pull request by tmichel**
## Rails

* Add support for some Rails 4 special methods (e.g. `before_action`)
* Fix highlighting when special methods are used as symbols

## Ruby

* Add pattern for constant definitions, apply `entity.name.type` so Go To Definition finds it
* Fix logical operators highlighting: `and`, `or`, `not` could be methods and in that case should not be highlighted.
* Add `prepend` as a special method. It is very similar to `include`.
* Add support for `%I` and `%i`: symbol arrays
* Remove `%S` and `%R` as they are not valid percent expressions
* Block variables can have modifiers (`*` and `&`), recognize these with the same scope as in methods.
* Add support for string formatting literals. E.g. `%s` or `%f`. Code copied for the Go syntax definition.

### CODE REVIEW

**Comment by wbond** [comment]
We'll definitely need to add some tests for these changes. See https://github.com/sublimehq/Packages#pull-requests and https://www.sublimetext.com/docs/3/syntax.html#Testing for more info.

**Comment by tmichel** [comment]
I'll add the tests for the modifications and the obvious ones that comes along. I think completely covering the Ruby & Rails syntax is out of scope in the pull request.

**Comment by wbond** [comment]
Yes, adding some basic tests plus targeted tests for these changes is all that is needed. No need to test all existing edge cases.

**Comment by tmichel** [comment]
@wbond I added tests and rebased it on `master` and incorporated the existing ruby tests.

File: Rails/Ruby on Rails.sublime-syntax [hunk 1]
```diff
@@ -87,10 +87,12 @@ contexts:
           pop: true
         - include: 'scope:source.ruby'
         - include: main
-    - match: \b(before_filter|skip_before_filter|skip_after_filter|after_filter|around_filter|filter|filter_parameter_logging|layout|require_dependency|render|render_action|render_text|render_file|render_template|render_nothing|render_component|render_without_layout|rescue_from|url_for|redirect_to|redirect_to_path|redirect_to_url|respond_to|helper|helper_method|model|service|observer|serialize|scaffold|verify|hide_action|append_view_path|prepend_view_path|view_paths)\b
+    - match: '\b(before_(filter|action)|skip_before_(filter|action)|skip_after_(filter|action)|after_(filter|action)|around_(filter|action)|filter|filter_parameter_logging|layout|require_dependency|render|render_action|render_text|render_file|render_template|render_nothing|render_component|render_without_layout|rescue_from|url_for|redirect_to|redirect_to_path|redirect_to_url|respond_to|helper|helper_method|model|observer|serialize|scaffold|verify|hide_action|append_view_path|prepend_view_path|view_paths)(?![?!:])\b'
```
Comment (wbond) [inline Rails/Ruby on Rails.sublime-syntax]: `(?![?!:])\b` will probably not function quite the way we want.

`\b` is a word boundary, so it won't trigger unless `?`, `!` or `:` are followed by a word character, or there is no symbol at the end of the method name.

Additionally, I presume the `(?![?!:])` is to handle Ruby 1.9 symbol keys, correct? If so, it should probably be: `\b(?![?!]?:)`. That will prevent matching the following: `after_filter!:`, `after_filter?:` and `after_filter:`. It will still match `after_filter :foo_bar`.

File: Ruby/Ruby.sublime-syntax [hunk 2]
```diff
@@ -440,7 +449,7 @@ contexts:
             0: punctuation.definition.string.end.ruby
           pop: true
         - include: regex_sub
-    - match: '%[QWSR]?\('
+    - match: '%[QWI]?\('
```
Comment (wbond) [inline Ruby/Ruby.sublime-syntax]: What triggered the change here?

File: Ruby/Ruby.sublime-syntax [hunk 3]
```diff
@@ -87,6 +87,11 @@ contexts:
         6: punctuation.separator.inheritance.ruby
         7: entity.other.inherited-class.module.third.ruby
         8: punctuation.separator.inheritance.ruby
+    - match: '([A-Z]\w*)(?=(,\s*[A-Z]\w*)*\s*=(?!\>))'
```
Comment (wbond) [inline Ruby/Ruby.sublime-syntax]: Let's add a test for this

File: Ruby/syntax_test_ruby.rb [hunk 4]
```diff
@@ -1,8 +1,50 @@
...
+    #                  ^^^^ variable.other.readwrite.instance.ruby
+    #                                ^ keyword.operator.arithmetic.ruby
+    #                                  ^ meta.variable.block.ruby
+    end
+  end
+
+  def my_method
+    str = "number %d" %  11
+                # ^^ string.quoted.double.ruby constant.other.placeholder.ruby
+    str = 'number %d' % 11
```
Comment (wbond) [inline Ruby/syntax_test_ruby.rb]: Can we add tests for a separator, field width and field number to ensure those are working properly and aren't broken via future changes?

**Comment by wbond** [comment]
Do your modifications handle the additions in #164 also?

File: Ruby/Ruby.sublime-syntax (hunk 2 above)
Comment (tmichel) [inline Ruby/Ruby.sublime-syntax]: `%S` and `%R` are not valid expressions.

```
irb(main):012:0> %S(test)
SyntaxError: (irb):12: unknown type of %string
%S(test)
   ^

irb(main):013:0> %R(test)
SyntaxError: (irb):13: unknown type of %string
%R(test)
   ^
```

File: Rails/Ruby on Rails.sublime-syntax (hunk 1 above)
Comment (tmichel) [inline Rails/Ruby on Rails.sublime-syntax]: You are correct. This is added to handle Ruby 1.9 symbol keys. 

I don't quite understand why but either version works. I added tests for recognizing `after_filter!:` and `after_filter?:` as a symbols.

File: Ruby/Ruby.sublime-syntax (hunk 3 above)
Comment (tmichel) [inline Ruby/Ruby.sublime-syntax]: Tests are in d28b28c.

File: Ruby/syntax_test_ruby.rb (hunk 4 above)
Comment (tmichel) [inline Ruby/syntax_test_ruby.rb]: I revised my previous version. Originally I copied it from the Go syntax definition but Ruby's format string is a bit different. So I adapted the regexp to better cover Ruby's format string.

The only thing not matched is named arguments.

```ruby
sprintf("%<foo>d : %<bar>f", { :foo => 1, :bar => 2 }) #=> 1 : 2.000000
sprintf("%{foo}f", { :foo => 1 }) # => "1f"
```

**Comment by tmichel** [comment]
> Do your modifications handle the additions in #164 also?

Yes. `xxx_filter` and `xxx_action` are both matched properly.

**Comment by wbond** [comment]
Excellent. Thank you for all of your work on this!
